cmd/git-sync/main.go

```diff
@@ -883,3 +893,23 @@ func setupGitAskPassURL(ctx context.Context) error {
 
 	return nil
 }
+
+
+func setupGitConfigs(ctx context.Context) error {
+	log.V(1).Info("adding git configs from flag")
+
+	for _, config := range strings.Split(*flGitConfigs, ",") {
```

Comment:
Are there any git-config keys or values that embed commas?  I know it's weird, but if a path had a comma in it, this would break.

This also applies to colons.  Simple string split probably is not enough.

A more robust approach would be a function that walks through the input string, one rune at a time and produces a `[]struct{k, v string}`, which you can then iterate over safely.  As you walk the string, any time you find a literal backslash, replace it with whatever comes next.

"\a\b\c" => "abc"
"foo\n" => "foon"
"hello\\there" => "hello\there"
"/weird/path/with/\,comma" => "/weird/path/with/,comma"

Or are we SURE that we never have to deal with this?

Reply:
Then we want a unit test :)

Reply:
As you pointed out, I believe I can't be sure that there won't be any commas and semicolons in these variables.

I'll change the function to go over rune at a time and escape based on backslashes.

The thing that concerns me is that you can actually put backslashes in gitconfig with double backslash `\\`

And if users want to put double backslash it would mean that they have to put `\\\\` with the function logic. 

`"\a\b\c"` => `"abc"`
`"foo\n"` => `"foon"`
`"hello\there"` => `"hellothere"`
`"/weird/path/with/\,comma"` => `"/weird/path/with/,comma"`

`"weird\\\\backslashid"` => `"weird\\backslashid"`

Does this sound fine?

Reply:
I'll definitely add a unit test to check this logic :)

Reply:
This got long.  I spent a lot of time playing with git-config, finding edge-case behaviors. 
 PLEASE double check me.

Looking at the git-config docs:

```
Section names are case-insensitive. Only alphanumeric characters, - and . are allowed in section names.
```

and 

```
The variable names are case-insensitive, allow only alphanumeric characters and -, and must start with an alphabetic character.
```

So keys can not hold comma or colon.  if you walk the string and parse the flag-string into a []string, delimited by "true" commas, you can take everything up to the first `:` as key (stripping leading/trailing whitespace) and everything after that (stripping leading/trailing whitespace) as value.  No need to escape `:` at all. 

Regarding whitespace on values:

```
Leading whitespaces after name =, ... and trailing whitespaces of the line are discarded unless they are enclosed in double quotes. Internal whitespaces within the value are retained verbatim.
````

To break the flag-string, though, you need to be smart.

Let's consider a sort of pathological case.  A key with a value that is a list, and another key with a simple value.

key1 = "k1"
value = "a:b, b:c, c:d"
key2 = "k2"
value = "v2"

Naive answer: `k1:a:b, b:c, c:d, k2:v2`

Obviously that doesn't work - it's ambiguous.

Option 1, escape commas: `k1:a:b\, b:c\, c:d, k2:v2`

Option 2 embed quotes: `k1:"a:b, b:c, c:d", k2:v2`

So I THINK we can say something like:

* The value is a series of comma-delimited key-value pairs.
* Each key must be a valid git-config compatible "<section>.<variable>" (see git-config docs)
* Each value must be a either a quoted string or an unquoted string
* Within all values, the following escapes are supported:
  - \n => <newline>
  - \t => <tab>
  - \" => "
  - \, => ,
  - \\ => \
* Within unquoted values, commas MUST be escaped.  Within quoted values, commas MAY be escaped, but are not required to be.
* Any other escape sequence is an error

When we build the string to pass to `git config` we DO NOT reformat it (e.g. no %q).  Git is weird here, and (empirically) it seems correct to pass literal newlines, which git will internally escape.

If there's demand for more escapes we can consider expanding it later (e.g. git-config supports \b, but I don't see why we should).

Now, keep in mind that to get these INTO git-sync, we need to pass them through a shell or YAML or something, so they may get escaped again.  That seems unavoidable.  In truth, I hope the vast vast majority of people just use simple tokens, buy we should be precise and code defensively. :)

We'll need a good test to cover the various ways one could try to break this.

Reply:
I've gone through the documentation as well.
I'll go and work on the codes for this and a solid test code as well.
btw thanks for the thorough review :)